SeedCone jet emulator

thesps opened the pull request:
This PR introduces the bit exact emulation of the seeded cone jet algorithm. It uses the classes, data types and constants from the correlator 'newfirmware' emulation.

Here are some performance plots on TTbar + 200 PU: https://cernbox.cern.ch/index.php/s/QzSAfx4hBzz1pBC

thesps committed:
Implement ap_fixed based emulator for L1SeedConePFJet.

thesps committed:
L1Trigger/Phase2L1ParticleFlow/plugins/L1SeedConePFJetProducer.cc

thesps committed:
Replace header, move pack to edm::Ptr

thesps committed:
Make L1SeedConePFJetEmulator standalone

thesps committed:
Add includes to L1SeedConePFJetEmulator for the standalone compile

schhibra committed:
use of calibrated-EM energy for HGCal EM-interpreted clusters

gpetruc committed:
Update EM ID storage in clusters

thesps committed:
Update SC jet emulator to use newfirmware/ objects, datatypes, constants

thesps committed:
Push rm of old emulator

gpetruc committed:
Make the parameter choosing the EM ID an enum, and turn it on by default

gpetruc commented on L1Trigger/Phase2L1ParticleFlow/plugins/L1SeedConePFJetProducer.cc:
```diff
@@ -109,14 +139,38 @@ void L1SeedConePFJetProducer::produce(edm::StreamID /*unused*/,
+  std::for_each(hwJets.begin(), hwJets.end(), [&](Jet jet){
+    edmJets.push_back(l1t::PFJet(jet.hwPt,
+                                 float(jet.hwEta) * l1ct::Scales::ETAPHI_LSB,
+                                 float(jet.hwPhi) * l1ct::Scales::ETAPHI_LSB));
```
Better `PFJet(jet.floatPt(), jet.floatEta(), jet.floatPhi(), /*mass=*/0., jet.intPt(), jet.intEta(), jet.intPhi())`

also, could be worth extending the PFJet class to store also the packed version as `uint64_t packed[2]` or `std::array<uint64_t,2>` (or `std::bitset<128>` , though from https://github.com/root-project/root/issues/6412 I take that ROOT isn't handling it great and anyway HLS doesn't support a good conversion from bitset to ap_uint)

gpetruc committed:
Update single-particle calibrations

gpetruc committed:
Update JECs for ak4 PF & Puppi

gpetruc committed:
Code-format

gpetruc committed:
Merge pull request #54 from gpetruc/EmInterp_L1PF_11_1_7_X_newfirmware

Bring EM interpretation to 11_1_7 branch

gpetruc commented on L1Trigger/Phase2L1ParticleFlow/plugins/L1SeedConePFJetProducer.cc:
```diff
@@ -109,14 +139,38 @@ void L1SeedConePFJetProducer::produce(edm::StreamID /*unused*/,
+  Config config(_debug, _coneSize, _nJets);
+  std::vector<Particle> particles = convertEDMToHW(work);
+  std::vector<Jet> jets = emulateEvent(particles, config);
+  return convertHWToEDM(jets);
```
In order to allow tracing back the PF cands, one thing you could do is:
 * in `convertEDMtoHW`, set also the [srcCand](https://github.com/thesps/cmssw/blob/L1PF_11_1_7_X_newfirmware_jetemulator/L1Trigger/Phase2L1ParticleFlow/src/newfirmware/dataformats/layer1_emulator.h#L108) to `edmParticle.get()` , and fill a `std::[unordered_]map<const l1t::PFCandidate *, edm::Ptr<l1t::PFCandidate>>` to remember the association
 *  in the emulator, return also the std::vector<Particle> of the constituents for each jet (either directly or perhaps making a l1ct::JetEmu class that extends l1ct::Jet class adding the vector of constituents?)
 * in `convertHWToEDM`, use the `srcCand` and map to get back the `edm::Ptr<l1t::PFCandidate>` and add the constituent to the jet.

gpetruc commented on L1Trigger/Phase2L1ParticleFlow/src/newfirmware/jetmet/L1SeedConePFJetEmulator.h:
```diff
@@ -0,0 +1,231 @@
+#include <numeric>
+#include <algorithm>
+
+namespace L1SCJetEmu {
```
This could become a class, with the Config passed in the constructor and the lookup table initialized only once in the constructor?

gpetruc commented on L1Trigger/Phase2L1ParticleFlow/src/newfirmware/jetmet/L1SeedConePFJetEmulator.h:
```diff
@@ -0,0 +1,231 @@
+
+  constexpr int floorlog2(int x) { return (x < 2) ? 0 : 1 + floorlog2(x / 2); }
+
+  constexpr int pow2(int x) { return x == 0 ? 1 : 2 * pow2(x - 1); }
```
why not just `return (1 << x);` ?

gpetruc commented on L1Trigger/Phase2L1ParticleFlow/src/newfirmware/jetmet/L1SeedConePFJetEmulator.h:
```diff
@@ -0,0 +1,231 @@
+  using namespace L1SCJetEmu;
+  Particle seed = parts.at(0);
+
+  // Fine unless we start using saturation, in which case order matters
```
our `pt_t` type does have saturation enabled, but if it's implemented correctly by the library I hope it still doesn't matter in which order we sum things: it's all _unsigned_ numbers and so if the total would above the ceiling it should saturate at some point in the partial sum and stay saturated

thesps replied:
Right, actually this comment is left over from the implementation before moving to the other newfirmware data types, and I didn't have saturation. But you're right, since it's unsigned the order doesn't matter after all.

gpetruc reviewed (commented):
Thanks. Some suggestions for improving it further are in the comments.

gpetruc committed:
Update version number for the dump

gpetruc committed:
Fix coordinates of muon input

gpetruc committed:
Bugfixes: fix muon coordinates, set refs in Puppi cands

gpetruc committed:
Better debugging info

gpetruc committed:
Merge pull request #55 from gpetruc/for_endcap_eg

Update EM ID information stored in emulator clusters

gpetruc committed:
Split e/gamma and PF-em clusters

gpetruc committed:
Set object multiplicities for endcaps & barrel (the latter from Dylan)

cerminar committed:
sync EG emulator and configuration

cerminar committed:
update parameter names for the EG emulator

gpetruc committed:
Merge pull request #57 from cerminar/for_endcap_eg_egintetg_v0

Update EG emulator and its configuration

gpetruc committed:
Update endcap regionizer, add EG preselection module

gpetruc committed:
code-format

gpetruc committed:
Merge pull request #58 from gpetruc/for_endcap_eg-emulator_cleanup

Update more emulators: regionizer, EG input selection

thesps committed:
Make a class for the emulator

thesps committed:
Add 128b jet word

gpetruc committed:
Merge pull request #59 from p2l1pfp/for_endcap_eg

Update E/gamma emulators and regionizer to support realistic layer1 endcap setup

dsrankin committed:
adding barrel regionizer emulator

thesps committed:
Retain access to the jet constituents from the emulator, implementing a map

thesps commented on L1Trigger/Phase2L1ParticleFlow/src/newfirmware/dataformats/puppi.h:
```diff
@@ -147,18 +147,20 @@ namespace l1ct {
       _pack_into_bits(ret, start, hwData);
       return ret;
     }
+    inline void initFromBits(const ap_uint<BITWIDTH> &src) {
+      unsigned int start = 0;
+      _unpack_from_bits(src, start, hwPt);
+      _unpack_from_bits(src, start, hwEta);
+      _unpack_from_bits(src, start, hwPhi);
+      _unpack_from_bits(src, start, hwId.bits);
+      _unpack_from_bits(src, start, hwData);
+    }
```
I added this since I needed to unpack a `PuppiObjEmu` from the packed bits rather than `PuppiObj`, and couldn't cast between. Perhaps there was a better way to implement it...

thesps commented:
I think I've addressed the comments:

- Emulator becomes a class with the config provided to the constructor
- Emulator jet constituents are tracked, using the method described with a map from the emulator objects to `edm::Ptr`s 
- PFJet has a member for the packed bits

cerminar committed:
fix index for EG ref creation when populating TkEM and TkEle collections

gpetruc committed:
Merge pull request #62 from cerminar/L1PF_11_1_7_X_newfirmware_egref_fix

Fix creation of reference to the standalone EG objects for TkEG objects

gpetruc committed:
Merge remote-tracking branch 'l1pf/L1PF_11_1_7_X_newfirmware' into drankincms-L1PF_11_1_7_X_newfirmware_TDRRegionizer

gpetruc committed:
code-format & code-checks

gpetruc committed:
Merge branch 'L1PF_11_1_7_X_newfirmware_jetemulator' of https://github.com/thesps/cmssw into thesps-L1PF_11_1_7_X_newfirmware_jetemulator

gpetruc committed:
Code cleanup for L1SeedConePFJetEmulator

gpetruc committed:
Synch includes with correlator-common

thesps committed:
Fix seed finding ordering in case of equal pT seeds in jet seededcone emulator
